### PR TITLE
[docs] Add PyTorch loaders article release

### PR DESCRIPTION
**Pull request by pablo-gar**
Images don't render in markdown, they are encoded for the [Myst parser](https://myst-parser.readthedocs.io/en/latest/syntax/images_and_figures.html#figures-images-with-captions)

### CODE REVIEW

File: docs/articles/2024/20240702-pytorch.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+:figwidth: 80%
+
+**Figure 4. Trial scVI training run with default parameters of the Census Pytorch loaders, highlighting increased speed of dense vs sparse data processing.** Training was done on X mouse cells for 1 epoch in X EC2 instance #TODO ask ebezzi for further details.
+```
+
+### Custom data encoders
+
+For maximum flexibility, users can provide custom encoders for the cell metadata enabling custom transformations or interactions between different metadata variables.
+
+To use custom encoders you need to instantiate the desired encoder via the `Encoder` (#TODO ebezzi to insert a link to the docs) class and pass it to the `encoders` parameter of the `ExperiementDataPipe`
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md:140]: ExperimentDataPipe
```suggestion
To use custom encoders you need to instantiate the desired encoder via the `Encoder` (#TODO ebezzi to insert a link to the docs) class and pass it to the `encoders` parameter of the `ExperimentDataPipe`
```

File: docs/articles/2024/20240702-pytorch.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+The single-cell expression data is encoded in numerical tensors, and for supervised training the cell metadata can be automatically transformed with a default encoder, or with custom user-defined encoders (see below).
+
+### Efficient memory usage with out-of-core training
+
+Thanks to the underlying backend of Census — TileDB-SOMA — the PyTorch loaders take advantage of incremental data materialization of fixed and small size to keep memory usage constant throughout training.
+
+In addition, data is eagerly fetched while batches go through training so that compute is never idle or waiting for data to be loaded. This feature is particularly useful when fetching Census data directly from the cloud.
+
+Memory usage is defined by the parameters #TODO ask ebezzi to fill out.
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md:80]:
```suggestion
Memory usage is defined by the parameters `soma_chunk_size` and `shuffle_chunk_count` - see below for a full description on how these should be tuned.
```

File: docs/articles/2024/20240702-pytorch.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+### Calibrated shuffling of observations (cells)
+
+Shuffling along efficient out-of-core data fetching is a challenge. In general, increasing randomness of shuffling leads to slower data fetching.
+
+In the first iteration of the loaders, shuffling was done through large blocks of data of user-defined size. This shuffling strategy led to non-random distribution of observations per training batch, becasue Census has a non-random data structure (observations from the same datasets are adjacent to one another) thus training loss was unstable (Figure 1).
+
+**Now we have implemented a scatter-gather approach**, whereby multiple chunks of data are fetched randomly from Census, then a number of chunks are concatenated into a block and all observations within the block are randomly shuffled. Adjusting the size and number of chunks per block leads to well-calibrated shuffling with stable training loss (Figure 2) while maintaining efficient data fetching (Figure 3).
+
+The balance between memory usage, efficiency, and level of randomness can be adjusted with the parameters #TODO ask ebezzi to fill in
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md:90]:
```suggestion
The balance between memory usage, efficiency, and level of randomness can be adjusted with the parameters `soma_chunk_size` and `shuffle_chunk_count`. Increasing `shuffle_chunk_count` will improve randomness, as more scattered chunks will be collected before the pool is randomized. Increasing `soma_chunk_size` will improve I/O efficiency while decreasing it will improve memory usage. We recommend a default of `soma_chunk_size=64, shuffle_chunk_count=2000` as we determined this configuration yields a good balance.
```

File: docs/articles/2024/20240702-pytorch.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+:figwidth: 80%
+
+**Figure 2. Training loss is well-calibrated with the current scatter-gather shuffling strategy.** Based on a trial scVI run on 250K Census cells.
+```
+
+### Increased training speed 
+
+We have made improvements to the loaders to reduce the amount of data transformations required from data fetching to model training. One such important change is to encode the expression data as a dense matrix immediately after the data is retrieved from disk/cloud. 
+
+In our benchmarks, we found that densifying data increases training speed ~3X while maintaining relatively constant memory usage (Figure 3). However we still allow users to decide whether to process the expression data in sparse or dense format via the #TODO ask ebezzi to include name of parameter.
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md]: The parameter is `method`, but I believe @ryan-williams wanted to change it?

File: docs/articles/2024/20240702-pytorch.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+:figwidth: 80%
+
+**Figure 2. Training loss is well-calibrated with the current scatter-gather shuffling strategy.** Based on a trial scVI run on 250K Census cells.
+```
+
+### Increased training speed
+
+We have made improvements to the loaders to reduce the amount of data transformations required from data fetching to model training. One such important change is to encode the expression data as a dense matrix immediately after the data is retrieved from disk/cloud.
+
+In our benchmarks, we found that densifying data increases training speed ~3X while maintaining relatively constant memory usage (Figure 3). However we still allow users to decide whether to process the expression data in sparse or dense format via the #TODO ask ebezzi to include name of parameter.
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md:109]:
```suggestion
In our benchmarks, we found that densifying data increases training speed ~3X while maintaining relatively constant memory usage (Figure 3). However we still allow users to decide whether to process the expression data in sparse or dense format via the `method` parameter.
```

File: docs/articles/2024/20240702-pytorch.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+We repeated the benchmark in Figure 3 in different conditions encompassing varying number of total cells and multiple epochs, please [follow this link for the full benchmark report and code.](https://github.com/ryan-williams/arrayloader-benchmarks).
+
+When comparing dense vs sparse processing in an end-to-end training exercise with scVI, we also observed slight increased speed with the dense approach and comparable memory usage to sparse processing (Figure 4). However in this full training example the differences were less substantial, highlighting that other model-specific factors during the training phase will contribute to memory and speed performance.
+
+```{figure} ./20240702-pytorch-fig-scvi.png
+:alt: Census scVI PyTorch run
+:align: center
+:figwidth: 80%
+
+**Figure 4. Trial scVI training run with default parameters of the Census Pytorch loaders, highlighting increased speed of dense vs sparse data processing.** Training was done on X mouse cells for 1 epoch in X EC2 instance #TODO ask ebezzi for further details.
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md:130]:
```suggestion
**Figure 4. Trial scVI training run with default parameters of the Census Pytorch loaders, highlighting increased speed of dense vs sparse data processing.** Training was done on 5684805 mouse cells for 1 epoch on a g4dn.16xlarge EC2 machine.
```

File: docs/articles/2024/20240702-pytorch.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+```
+
+### Custom data encoders
+
+For maximum flexibility, users can provide custom encoders for the cell metadata enabling custom transformations or interactions between different metadata variables.
+
+To use custom encoders you need to instantiate the desired encoder via the `Encoder` (#TODO ebezzi to insert a link to the docs) class and pass it to the `encoders` parameter of the `ExperimentDataPipe`.
```
Comment (ebezzi) [inline docs/articles/2024/20240702-pytorch.md:137]:
```suggestion
To use custom encoders you need to instantiate the desired encoder via the [Encoder](https://chanzuckerberg.github.io/cellxgene-census/_autosummary/cellxgene_census.experimental.ml.encoders.Encoder.html#cellxgene_census.experimental.ml.encoders.Encoder) class and pass it to the `encoders` parameter of the `ExperimentDataPipe`.
```

Note that this link will be broken until we release.

File: docs/articles/2024/20240709-pytorch.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
+# First stable iteration of Census (SOMA) PyTorch loaders
+
+*Published:* *July 9th, 2024*
+
+*By:* *[Emanuele Bezzi](mailto:ebezzi@chanzuckerberg.com), [Pablo Garcia-Nieto](mailto:pgarcia-nieto@chanzuckerberg.com), [Prathap Sridharan](mailto:psridharan@chanzuckerberg.com), [Ryan Williams](mailto:pgarcia-nieto@chanzuckerberg.com)*
```
Comment (ebezzi) [inline docs/articles/2024/20240709-pytorch.md:5]: Ryan's email is wrong. Worth checking if they're ok with adding the email here though?

**Review by ebezzi — APPROVED** [review]

**Comment by ryan-williams** [comment]
Thanks for this! One note: the wrong circle is labeled "default" in [docs/articles/2024/20240709-pytorch-fig-benchmark.png](https://github.com/chanzuckerberg/cellxgene-census/pull/1214/files#diff-7f213570e373be149d6e854425e7d4359b642b6b1eef94124c23e95c21be58cf):

image

From [the interactive plot](https://rw-tdb.s3-us-west-2.amazonaws.com/arrayloader-benchmarks/notebooks/data-loader/:138_4096/speed_vs_mem_1.html), here's a gif showing the 2 data points corresponding to 2048 chunks of size 64 (one from a g4dn.4xlarge, one from a g4dn.8xlarge):

default

<details>
<summary>
The circle currently labeled "default" actually corresponds to 1024 chunks of size 128.
</summary>

default-1024-4x
</details>

I'm not sure how significant the speed difference is between the g4dn.{4x,8x}large nodes, with the default configuration, given N=1. I can generate a few more samples, if you like.

File: docs/articles/2024/20240709-pytorch.md (hunk 8 above)
Comment (ryan-williams) [reply docs/articles/2024/20240709-pytorch.md:5]: Thanks, I saw that #1228 addresses this 🙏
